src/storage/src/hummock/sstable_store.rs

```diff
@@ -926,27 +986,49 @@ impl BlockStream {
         }
 
         let block_meta = &self.block_metas[self.block_idx];
-        let mut buffer = vec![0; block_meta.len as usize];
         fail_point!("stream_read_err", |_| Err(HummockError::object_io_error(
             ObjectError::internal("stream read error")
         )));
-
-        let bytes_read = self
-            .byte_stream
-            .read_bytes(&mut buffer[..])
-            .await
-            .map_err(|e| HummockError::object_io_error(ObjectError::internal(e)))?;
-
-        if bytes_read != block_meta.len as usize {
-            return Err(HummockError::decode_error(ObjectError::internal(format!(
-                "unexpected number of bytes: expected: {} read: {}",
-                block_meta.len, bytes_read
-            ))));
-        }
+        let offset = block_meta.offset as usize;
+        let end = offset + block_meta.len as usize;
+        let data = if end > self.file_pos + self.buff.len() {
+            let mut read_buf = BytesMut::with_capacity(block_meta.len as usize);
+            let start_pos = if self.buff_offset < self.buff.len() {
+                assert!(offset < self.file_pos + self.buff.len());
+                read_buf.extend_from_slice(&self.buff[self.buff_offset..]);
+                self.buff.len() - self.buff_offset
+            } else {
+                0
+            };
+            let mut rest = block_meta.len as usize - start_pos;
+            self.buff_offset = 0;
+            self.file_pos += self.buff.len();
+            while rest > 0 {
+                let next_packet = self
+                    .byte_stream
+                    .read_bytes()
+                    .await
+                    .unwrap_or_else(|| Err(ObjectError::internal("read unexpected EOF")))?;
+                let read_len = std::cmp::min(next_packet.len(), rest);
+                read_buf.extend_from_slice(&next_packet[..read_len]);
+                rest -= read_len;
+                if rest == 0 {
+                    self.buff_offset = read_len;
```

Comment:
I am a bit confused. What does `file_pos` represent? I though it is the current offset of the SST file but it seems not because we don't update `file_pos` here.

Reply:
it represents the postition of left bound of buffer. We will update `file_pos` every time we read a new buffer.

Reply:
This code has been refactor and I removed the variable `file_pos`